ast+topdown: Add `graphql.is_schema_valid` builtin.

philipaconrad opened the pull request:
This commit adds a new GraphQL builtin for validating GraphQL schemas, applying stronger validation rules than what the current GraphQL parsing builtins apply by default.

Fixes: #5125

anderseknert commented on test/cases/testdata/graphql/test-graphql-is-schema-valid.yaml:
```diff
@@ -0,0 +1,115 @@
+# own test suite, as it provides a fairly comprehensive set of good/degenerate
+# test cases, which we want to make sure to react correctly to.
+# See: https://github.com/vektah/gqlparser/blob/master/validator/validator_test.yml
+cases:
```
Only happy path tests here, AFAICS.

philipaconrad replied:
I think the reason for that is because the original test suite I borrowed from for the `graphql.is_valid` builtin was just happy-path tests too. I can fix that though!

philipaconrad replied:
Added a stack of tests, adapted from the `gqlparser` test suite. Marking as resolved. :+1:

srenatus replied:
Committed, too? I can't find it 😅 but it's early, probably just not seeing it.

srenatus replied:
Found it. Ignore me.

anderseknert reviewed (commented):
Some tests for the error conditions would be great, but besides that, looks good to me!

srenatus commented on topdown/graphql.go:
```diff
@@ -453,10 +453,38 @@ func builtinGraphQLIsValid(_ BuiltinContext, operands []*ast.Term, iter func(*as
+
+	// Validate the schema, erroring if there's an issue.
+	_, err = convertSchema(schemaDoc)
+	if err != nil {
+		return iter(ast.BooleanTerm(false))
+	}
+
+	// If we got this far, the GraphQL query passed validation.
+	return iter(ast.BooleanTerm(true))
```
[golf] (let's not do that 😄)
```suggestion
	return iter(ast.BooleanTerm(err != nil))
```

philipaconrad replied:
Rolled in the suggestion here. Marking as resolved. :+1:

srenatus reviewed (dismissed):
Nothing to add over @anderseknert's review. 👍

philipaconrad commented:
I renamed the builtin to match what we've done over on PR #5306, with `net.cidr_is_valid`. I've also added additional tests (adapted from `gqlparser`'s parser schema tests), and have preemptively squash-rebased the commits to get this PR ready for merging.
